runtime-sdk/src/context.rs

```diff
@@ -172,9 +171,6 @@ pub struct DispatchContext<'a> {
     pub(crate) io_ctx: Arc<IoContext>,
     pub(crate) logger: slog::Logger,
 
-    /// The runtime's methods, in case you need to look them up for some reason.
-    pub(crate) methods: &'a MethodRegistry,
```

Comment:
This may actually be useful for thing like calling transactions on behalf of a smart contract etc.

Reply:
been thinking about this, and there are two ideas that make this previous attempt look very hacky. but this previous attempt was the result of some seemingly incidental possibly temporary opportunity.

currently there's this awkward imbalance where the method registry use function pointers and is type-agnostic of the runtime, while other module-provided functionality (e.g. authenticate tx) uses static dispatch and thus depends on runtime type. overall, the dispatcher needs both, so it depends on the runtime type. we went with this give-them-the-method-registry-only ~hack which (i) was all we needed before tx auth gas and (ii) worked because we used function pointers for those methods.

the goal for these two ideas is to make dispatcher access more general, so we can additionally call the other things, such as a hypothetical new pre-transaction callback. we'd do this by passing the entire dispatcher into the method handler.

1. give handlers a dispatcher type parameter. this would let us pass a Dispatcher<R> into it without having to make the module as a whole know about R, which I think wouldn't work because R::Modules needs to contain the module.
2. make the whole dispatcher use function pointers and not depend on R. or other dynamic dispatch. then we'll be able to pass the dispatcher reference into the handler

Reply:
Yeah, agreed on liking the static dispatch. Should be doable, just wondering how a bunch of `match` statements would compare to a map lookup? Probably doesn't matter unless you have lots of methods.

Reply:
dispatch from within methods is reinstated in #147. and with that, I've moved gas to be encapsulated in the core module again. yay